docs/charms.reactive.decorators.rst

```diff
@@ -3,6 +3,38 @@ charms.reactive.decorators
 
 .. rubric:: Summary
 
+These decorators turn a regular function into a reactive handler so that it will
+be invoked if the conditions in the decorators are met. These decorators are
+also called the preconditions of the handlers. If a handler has multiple
+decorators, they will be ANDed together: the handler will only be invoked if
+all of the individual decorators match.
+
+Handlers are currently re-invoked for every hook execution, even if their
+predicate flags have not changed. However, this may well change in the future,
+and it is highly recommended that you not rely on this behavior. If you need to
+ensure that a particular handler runs again, you should set or remove one of its
+predicate flags.
+
+Regular handlers should not accept any arguments. When a handler needs to use a
+(relationship) :class:`~charms.reactive.altrelations.Endpoint`, it can access
+the endpoint object via the :data:`~charms.reactive.helpers.context` namespace.
+The only exceptions to this are enpoint handlers, handlers that are instance
+methods of :class:`~charms.reactive.altrelations.Endpoint`: they get the
+endpoint object as the `self` argment.
+
+If a handler needs to use an older interface using the legacy
+:class:`~charms.reactive.relations.RelationBase`, or if you're uncertain which
+is used by an interface layer, you can also obtain an instance using
+:func:`~charms.reactive.relations.relation_from_flag`.
+
+For backwards compatibility, some decorators can pass instances of these
```

Comment:
How do we handle the case when a handler doesn't accept instances but the decorator wants to pass one? (a "new" handler reacting to a flag of an "old" interface)

Reply:
I added introspection logic to the invocation code so that, if the handler doesn't take any arguments, it will do the right thing and not pass them.  You'd then have to use `relation_factory` or `context.endpoints` to access them, depending on how they're implemented (ugh).  Though, the decorators will also pass the args through for `EndpointInterface` flags as well, so we'll probably end up having to continue to do that until everything is converted and then we'll still have the huge amount of inertia to move away (ugh).